sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala

```diff
@@ -840,6 +842,10 @@ object TypeChecks {
     areTimestampsSupported(ZoneId.systemDefault()) &&
       areTimestampsSupported(SQLConf.get.sessionLocalTimeZone)
   }
+
+  def isTimezoneSensitiveType(dataType: DataType): Boolean = {
+    dataType == DateType || dataType == TimestampType
```

Comment:
I think yes according to [the timezone requirement check ](https://github.com/apache/spark/blob/v3.1.1/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala#L108)in Cast.
```
  /**
   * Return true if we need to use the `timeZone` information casting `from` type to `to` type.
   * The patterns matched reflect the current implementation in the Cast node.
   * c.f. usage of `timeZone` in:
   * * Cast.castToString
   * * Cast.castToDate
   * * Cast.castToTimestamp
   */
  def needsTimeZone(from: DataType, to: DataType): Boolean = (from, to) match {
    case (StringType, TimestampType | DateType) => true
    case (TimestampType | DateType, StringType) => true
    case (DateType, TimestampType) => true
    case (TimestampType, DateType) => true
    case (ArrayType(fromType, _), ArrayType(toType, _)) => needsTimeZone(fromType, toType)
    case (MapType(fromKey, fromValue, _), MapType(toKey, toValue, _)) =>
      needsTimeZone(fromKey, toKey) || needsTimeZone(fromValue, toValue)
    case (StructType(fromFields), StructType(toFields)) =>
      fromFields.length == toFields.length &&
        fromFields.zip(toFields).exists {
          case (fromField, toField) =>
            needsTimeZone(fromField.dataType, toField.dataType)
        }
    case _ => false
  }
```
Besides, `DateAddInterval` for date revelant calculation, is also a TimeZoneAwareExpression. I checked the `eval` [code](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala#L1700), it indeed passes in the `zoneId`.
 ```
      val startTs = DateTimeUtils.daysToMicros(start.asInstanceOf[Int], zoneId)
      val resultTs = DateTimeUtils.timestampAddInterval(
        startTs, itvl.months, itvl.days, itvl.microseconds, zoneId)
      DateTimeUtils.microsToDays(resultTs, zoneId)
```

Reply:
I did a little spelunking and 3 years ago https://issues.apache.org/jira/browse/SPARK-35581 removed zone ID from being used as a part of the cast. This went into 3.2.0. So anything in 3.1.X uses the time zone when doing a cast from string to date. Technically even then the time zone was only used for for special cases that were removed as a part of the Spark issue.  So it is a bug in Spark. We probably should put up a small PR to fix the issue in Spark. @firestarman if you want to do it the go ahead. If not I'll put up a PR.

Reply:
Hi @revans2, please help file a PR for that.

Things become more compplicated than I thought.
The timezone tagging in type checks may also be incorrect for some cases. For example, if casting a `Long` to `Timestamp`, it will fallback on CPU, but timezone is actually not used. 
We may need some refactor for timezone tagging and check if it is needed case by case. This appears to be a lot of work.

Reply:
Updated, I choose to do timzone tagging for `Cast` only when its input or ouput has `TimestampType`.

Reply:
https://issues.apache.org/jira/browse/SPARK-42898 Will try to get a PR up soon-ish